### PR TITLE
fix(rct): simplify error handling in run_exp

### LINKED ISSUE #107: Actors hang and resources not released when MLflow runs not marked FAILED before os._exit(1)
## Problem

After a 2-day full sweep on the GPU server, the Ray autoscaler started repeatedly warning:

```
Warning: The following resource request cannot be scheduled right now: {'CPU': 5.0, 'GPU': 1.0}.
This is likely due to all cluster resources being claimed by actors.
```

The sweep was triggered **before** the following fix was added:

```python
if logger is not None:
    logger.experiment.set_terminated(logger.run_id, status="FAILED")
```

However, `os._exit(1)` **was** already present when the run started.

## Root Cause

When a training run fails and `os._exit(1)` is called, MLflow runs are left in a `RUNNING` (or unknown) state because `set_terminated(..., status="FAILED")` was not being called beforehand. This means:

- MLflow has no record that these runs failed
- Actors may hang or fail to release GPU/CPU resources properly
- Pending tasks accumulate (100+ tasks waiting for `{'CPU': 5.0, 'GPU': 1.0}`) with no way to proceed

## Fix

The fix (partially in place) is to call `set_terminated` before `os._exit(1)`:

```python
if logger is not None:
    logger.experiment.set_terminated(logger.run_id, status="FAILED")
os._exit(1)
```

This ensures MLflow runs are properly closed as FAILED before the process is forcefully terminated.

## Impact

- Long sweeps can get silently stuck after failures, wasting GPU server time
- Difficult to diagnose because the resources appear claimed but no work is progressing

### LINKED ISSUE #130: Non-OOM exceptions leave MLflow runs stuck in RUNNING (stale)
## Bug

MLflow runs that fail with non-OOM exceptions are never marked as `FAILED`, leaving them permanently stuck in `RUNNING` status ("stale" runs).

## Root cause

In `stgym/rct/run.py`, the `set_terminated(..., status="FAILED")` call is gated behind `isinstance(e, torch.cuda.OutOfMemoryError)` (added in #99). For any other exception (NaN loss, shape mismatch, data loading error, etc.), the exception is caught and logged to an artifact, but the MLflow run is never explicitly ended — so it stays `RUNNING` on the server indefinitely.

This affects both the regular train/val/test path (line ~123) and the k-fold path (line ~176).

```python
except Exception as e:
    log_training_error(e, logger)
    if isinstance(e, torch.cuda.OutOfMemoryError):  # ← only OOM
        if logger is not None:
            logger.experiment.set_terminated(logger.run_id, status="FAILED")
        os._exit(1)
    # non-OOM exceptions: run stays RUNNING forever
```

## Observed impact

`sweep_status.py` reports stale runs even though #99 was expected to eliminate them:

```
dimension      task   ...  active  stale
model.act graph-clf  ...       4     10
```

## Proposed fix

Move `set_terminated(..., status="FAILED")` before the OOM check so it applies to **all** exceptions.

### PR DESCRIPTION
**Pull request by xiaohan2012**
## Problem

OOM-specific error handling in `run_exp` called `set_terminated()` and `os._exit(1)` only for `torch.cuda.OutOfMemoryError`, leaving all other exceptions silently swallowed — MLflow runs stuck in `RUNNING` state and callers receiving `True` despite failure.

Fixes #130
Fixes #107

## Solution

Both special-case behaviours are now redundant:
- `set_terminated(..., "FAILED")` — PyTorch Lightning's `MLFlowLogger.finalize("failed")` already calls this for **all** exceptions before they propagate out of `train()`.
- `os._exit(1)` — `max_calls=1` on the Ray remote already recycles the worker process after every task, so the GPU slot is always released.

Replace both `except` blocks with a plain `raise`, so all exceptions propagate correctly.

## Changes

- `stgym/rct/run.py` — remove OOM gate and `os._exit(1)` from both the regular and k-fold except blocks; remove unused `import os`
- `tests/rct/test_run.py` — delete `test_oom_kills_worker_process` (tests removed behaviour); update `test_error_handling` to expect `RuntimeError` raised instead of swallowed

## Testing

- [x] `pre-commit run --all-files` passes
- [x] `pytest tests/ -m 'not slow'` passes (7/7 in `tests/rct/test_run.py`)
- [ ] Run a full sweep on cyy2 and verify `sweep_status.py` reports zero stale runs

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

**Comment by xiaohan2012** [comment]
## Sweep validation update

Ran a full design-dimension sweep (experiment `sweep-all-20260420-190614`, ID 12) on cyy2 after merging this branch. Results after ~8h:

- **994 runs** across 14 dimensions
- **936 finished** / 52 failed / 0 active
- **Stale runs: 0** ✅

The `stale` column is 0 for every dimension — failed runs are now properly terminated by Lightning's `MLFlowLogger.finalize("failed")` and never left stuck in `RUNNING`.

2 dimensions still pending (`data_loader.radius_ratio`, `model.num_mp_layers`) but the fix is confirmed working. Checking off the sweep test item.

**Comment by xiaohan2012** [comment]
The previous throughput was around 60 runs per hour. Now it doubles. I think this is related to the memory accumulation fix #143
